### ISSUE TITLE
 MD5SUM updating at every execution

### ISSUE THREAD

**Issue by doakd** [liquibase/liquibase#3269]
## Environment

**Liquibase Version**: Community 4.14

**Liquibase Integration & Version**: CLI

**Liquibase Extension(s) & Version**: Oracle 4.14, Teradata 4.14, neo4j 4.14, netezza 4.14 (dbeaver extension)

**Database Vendor & Version**: Oracle 12.1.0.2

**Operating System Type & Version**: Linux 5.4.0-1084-aws #91~18.04.1-Ubuntu SMP x86_64 GNU/Linux

**Infrastructure Type/Provider**: AWS

## Description

I have a customer who has a large number of changesets that get the MD5SUM updated in databasechangelog at the start of every deployment, even though the changesets have not been modified and the changesets are not actually executed.  The Liquibase output looks like this:

```
[2022-09-02 15:46:24] INFO [liquibase.executor] UPDATE TST4CIS.DATABASECHANGELOG SET MD5SUM = '8:cf74d98bf9923254d88d77efba90e1ec' WHERE ID = 'ads_cisd015.pks' AND AUTHOR = 'xxx' AND FILENAME = 'Packages/db-changelog-1.xml'
[2022-09-02 15:46:24] INFO [liquibase.executor] 1 row(s) affected
[2022-09-02 15:46:24] INFO [liquibase.executor] UPDATE TST4CIS.DATABASECHANGELOG SET MD5SUM = '8:a1d70adee4b5b830d4b5ab103c371548' WHERE ID = 'ads_cisd109.pks' AND AUTHOR = 'xxx' AND FILENAME = 'Packages/db-changelog-1.xml'
[2022-09-02 15:46:24] INFO [liquibase.executor] 1 row(s) affected
[2022-09-02 15:46:24] INFO [liquibase.executor] UPDATE TST4CIS.DATABASECHANGELOG SET MD5SUM = '8:859cb9b9b4174f5322329ee434232c55' WHERE ID = 'bobfixtca.pkb' AND AUTHOR = 'xxx' AND FILENAME = 'Packages/db-changelog-1.xml'
[2022-09-02 15:46:24] INFO [liquibase.executor] 1 row(s) affected
```

It's strange because it's not all of his changesets, only the changesets that contain PL/SQL (packages, procedures, trigggers, etc) , executed via sqlFile, and the files have .pkb, .pks, .prc, and .sql extensions.  

This seems to have started after two changes we made:
1. Upgraded to 4.12
2. Added sql-log-level=INFO

This is only occurring with this one customer (we have 100s) so I'm assuming it's something about his changelogs that is causing this, but I don't see anything obvious.

**Comment by kataggart** [comment]
@doakd Would it be possible for you to share the changelogs that are leading to this behavior? I understand if you prefer to keep those private; you can email them to me at ktaggart@liquibase.com (with any private info redacted/removed of course).

Also would be helpful to see one that is not having this issue.

What version did you update from to 4.12 when you started to see this issue?

Thanks!

**Comment by doakd** [comment]
Files have been sent.  

Looks like the issue is only occurring when sqlFile is used in xml (that I'm aware of anyway).  Does not occur when .sql files are used.

We were previously using 4.9.1.

**Comment by doakd** [comment]
Is there an update on this issue?  

This is also preventing runOnChange from working properly, so the changesets have to be converted to runalways to a have a modification to the PL/SQL executed.

Thanks.

**Comment by tabbyf00** [comment]
Hey Daryl.  Nathan and Fede are looking into it.  We'll post an update as soon as we have one. Thanks for following up!

**Comment by FBurguer** [comment]
Hi! we have been looking at this issue and we are unable to find out if this issue is a bug on our code or is an error on your end. Is it possible that your changing MD5SUM on one part of your code or inside of one of your changelogs? The other possible root source can be that your jumping between liquibase versions, but it doesnt make sense just to be in some of the changelogs and not in all of them, maybe the job that its executing these especific changesets has a different liquibase version than the rest?

**Comment by nvoxland** [comment]
As a little more background: the only times you should get a `UPDATE TST4CIS.DATABASECHANGELOG SET MD5SUM = '8:cf74d98bf9923254d88d77efba90e1ec' WHERE ID ...` message is when the md5sum value is either NULL or starting with a different "version" (the number before the colon).

If it was a problem where the checksum keeps being computed differently each time, you'd see the md5sum updated as part of a larger updates statement that also sets the description, marks it as reran, etc.

The only time you'd have a difference in the version is when you are going between liquibase versions where we had a large algorithm change. These are few and far between, but would only be happening if you ran different liquibase versions on different runs. Is that something you are doing?

The only time the the md5sum would be null is if you ran changelogsync or had some other external process that sets the checksum to be null.  Any chance you're doing something like that somewhere?

In either of those two cases, the fact that it's only a few of the checksums that have a problem is interesting. Running changelogsync or running update with an old version of liquibase will update ALL the checksums, not just some. So if there is something outside of each liquibase update changing those, they are being ran against a child changelog instead of master? Or something manually that's just targettting those changesets?

**Comment by doakd** [comment]
The MD5SUM values are not NULL.
The MD5SUM values are not being updated outside of Liquibase.
changelogsync has not been run.
We are only using version 4.14.

Only files using sqlFile from an xml changelog are impacted.  The .sql files that are using include from xml are fine.

**Comment by CGarces** [comment]
Same issue.

This is the error that I see in the logs:

`config/liquibase/changelog/00000000000000_initial_schema.xml::00000000000001::jhipster was: 8:8bacaa5647e7cc5246fc7d5c5d4d7cde but is now: 8:3fa67a5aeb6197bd6ea8fb1df5ee943f`

The file 00000000000000_initial_schema.xml not has changes since some time ago, and I'm confident that my database is fine.

so I made:

```sql
update DATABASECHANGELOG set MD5SUM='8:3fa67a5aeb6197bd6ea8fb1df5ee943f' 
where MD5SUM='8:8bacaa5647e7cc5246fc7d5c5d4d7cde';
```

After restart the server (only restart, without any change) I get a different error.

`config/liquibase/changelog/00000000000000_initial_schema.xml::00000000000001::jhipster was: 8:3fa67a5aeb6197bd6ea8fb1df5ee943f but is now: 8:68f03dcc3c4ecfeba73f36c52d3cb4b4`

So MD5 changes after some deploys/restart.

I have fixed my problem with

`update DATABASECHANGELOG set md5sum=null`

But I don't understand the root cause.

My App is running on AWS Elastic Beanstalk with Tomcat 8.5 with Corretto 11 running on 64bit Amazon Linux 2/4.2.16

**Comment by nvoxland** [comment]
Are you able to enable fine-level logging which should output something like:

```
FINE [liquibase.util] Computed checksum for createTable:[
    columns=[
        [
            name="id"
            type="int"
        ]
    ]
    tableName="test_table2"
] as 0c3ee0cfd63d1fc0afeb2418835cba2a
``` 
for `config/liquibase/changelog/00000000000000_initial_schema.xml::00000000000001::jhipster` ?

That will give us the actual string being checksumed and we should be able to see a difference between the two runs in those strings the checksum is based off of.

**Comment by CGarces** [comment]
I'll try to reproduce the error, but is a production environment, it will take some time.

**Comment by ic185037** [comment]
Hi @nvoxland and all,

Thanks for keeping alive this issue.

We have a similar issue. In our case, we moved our environment from 4.4.3 to 4.16.1 and checksums calculated with the new version are different, despite of not doing any change to the changeSet (and regardless the fact that both algorithm versions are the same, version 8).

`1 changesets check sum
          update/services/messaging/messaging_tables.xml::1559570706::mw185149 was: 8:e1aa84a7a58273fe5908ad86a589612d but is now: 8:dce471b939276f17937a806ee31e276b`

We have also noticed that even the calculate-checksum generates a totally new checksum value for the same changeSet. That is, I am having at least 3 different checksum values

`liquibase calculate-checksum update/services/messaging/messaging_tables.xml::1559570706::mw185149 --defaultsFile=./cxthcs.onprem.oracle.user.properties --changeLogFile=./update/cxthcs_objects.xml                     
####################################################
 LOGO HERE
####################################################
Starting Liquibase at 09:43:47 (version 4.16.1 #4594 built at 2022-09-14 15:27+0000)
Liquibase Version: 4.16.1
Liquibase Community 4.16.1 by Liquibase
8:f8815ddd856f873e8dd36743b09635ea
Liquibase command 'calculate-checksum' was executed successfully.`

The issue is not limited to the Oracle environment (this issue was tagged with the DBOracle label) but also in SQL Server. Also, it doesn't seem related to a specific ChangeType but to the Liquibase version change. However, checksums from the calculate-checksums command between version 4.16.1 and 4.17.1 are the same. 

As a workaround, we are clearing the checksums during the version upgrade, but a proper solution is truly appreciated.

Looking forward to hearing from you.

Thanks!

**Comment by nvoxland** [comment]
We're still having a hard time finding a way to reproduce this to understand why it's happening. @ic185037 are you able to enable fine level logging to get the output like:

```
FINE [liquibase.util] Computed checksum for createTable:[
    columns=[
        [
            name="id"
            type="int"
        ]
    ]
    tableName="test_table2"
] as 0c3ee0cfd63d1fc0afeb2418835cba2a
```
for times it ends up computing different checksums? That's what we need to really understand what's going on

**Comment by ic185037** [comment]
Hi @nvoxland 

Thanks for your quick reply

Please find below the information you asked me:

Checksum information by the createTable changeSet executed with 4.4.3

`[2022-11-11 06:34:49] FINE [liquibase.util] Computed checksum for createTable:[
    columns=[
        [
            [
                nullable="false"
                primaryKey="true"
                primaryKeyName="PK_cs_guaranteed_delivery"
            ]
            autoIncrement="true"
            name="internal_id"
            type="bigint"
        ],
        [
            [
                nullable="false"
            ]
            name="message_header"
            type="nclob"
        ],
        [
            [
                nullable="false"
         ... [truncated in log] as 2ee3b0b9cebc0f4dc24d5293b30dad26
[2022-11-11 06:34:49] FINE [liquibase.util] Computed checksum for replace:[
    applicableDbms=[
        oracle,
        mssql
    ]
    applyToRollback="false"
    contexts="non-iot"
    replace=", CONSTRAINT PK_cs_guaranteed_delivery PRIMARY KEY (internal_id)"
    with=""
] as d0e8ec5e3b5380571154dfb483e6080a
[2022-11-11 06:34:49] FINE [liquibase.util] Computed checksum for regExpReplace:[
    applicableDbms=[
        oracle,
        mssql
    ]
    applyToRollback="true"
    replace="((?:[A-Z]+)_(?:DATA|INDX)_(?:FG|DF))"
    with="CXTHCS_$1"
] as 662da3b53a7d2ec55863e896369220ea
[2022-11-11 06:34:49] FINE [liquibase.util] Computed checksum for 8:2ee3b0b9cebc0f4dc24d5293b30dad26:8:d0e8ec5e3b5380571154dfb483e6080a;8:662da3b53a7d2ec55863e896369220ea; as e1aa84a7a58273fe5908ad86a589612d`

Checksum information by the createTable changeSet executed with 4.16.1

`[2022-11-11 07:45:28] FINE [liquibase.util] Computed checksum for createTable:[
    columns=[
        [
            [
                nullable="false"
                primaryKey="true"
                primaryKeyName="PK_cs_guaranteed_delivery"
            ]
            autoIncrement="true"
            name="internal_id"
            type="bigint"
        ],
        [
            [
                nullable="false"
            ]
            name="message_header"
            type="nclob"
        ],
        [
            [
                nullable="false"
         ... [truncated in log] as 2ee3b0b9cebc0f4dc24d5293b30dad26
[2022-11-11 07:45:28] FINE [liquibase.util] Computed checksum for replace:[
    applicableDbms=[
        oracle,
        mssql
    ]
    applyToRollback="false"
    contextFilter="non-iot"
    replace=", CONSTRAINT PK_cs_guaranteed_delivery PRIMARY KEY (internal_id)"
    with=""
] as f60940602dcd93a68b2b4a93da27ac35
[2022-11-11 07:45:28] FINE [liquibase.util] Computed checksum for regExpReplace:[
    applicableDbms=[
        oracle,
        mssql
    ]
    applyToRollback="true"
    replace="((?:[A-Z]+)_(?:DATA|INDX)_(?:FG|DF))"
    with="CXTHCS_$1"
] as 662da3b53a7d2ec55863e896369220ea
[2022-11-11 07:45:28] FINE [liquibase.util] Computed checksum for 8:2ee3b0b9cebc0f4dc24d5293b30dad26:8:f60940602dcd93a68b2b4a93da27ac35;8:662da3b53a7d2ec55863e896369220ea; as dce471b939276f17937a806ee31e276b
[2022-11-11 07:45:28] FINE [liquibase.executor] Release Database Lock
[2022-11-11 07:45:28] FINE [liquibase.executor] Executing with the 'jdbc' executor
[2022-11-11 07:45:28] INFO [liquibase.lockservice] Successfully released change log lock
[2022-11-11 07:45:28] FINE [liquibase.configuration] No configuration value for liquibase.outputLineSeparator found
[2022-11-11 07:45:28] FINE [liquibase.configuration] Configuration liquibase.outputLineSeparator is using the default value of 

[2022-11-11 07:45:28] SEVERE [liquibase.integration] Validation Failed:
     1 changesets check sum
          update/services/messaging/messaging_tables.xml::1559570706::mw185149 was: 8:e1aa84a7a58273fe5908ad86a589612d but is now: 8:dce471b939276f17937a806ee31e276b
liquibase.exception.CommandExecutionException: liquibase.exception.ValidationFailedException: Validation Failed:
     1 changesets check sum
          update/services/messaging/messaging_tables.xml::1559570706::mw185149 was: 8:e1aa84a7a58273fe5908ad86a589612d but is now: 8:dce471b939276f17937a806ee31e276b

	at liquibase.command.CommandScope.execute(CommandScope.java:179)
	at liquibase.integration.commandline.CommandRunner.call(CommandRunner.java:51)
	at liquibase.integration.commandline.CommandRunner.call(CommandRunner.java:21)
	at picocli.CommandLine.executeUserObject(CommandLine.java:1953)
	at picocli.CommandLine.access$1300(CommandLine.java:145)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2358)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2352)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2314)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2316)
	at picocli.CommandLine.execute(CommandLine.java:2078)
	at liquibase.integration.commandline.LiquibaseCommandLine.lambda$execute$1(LiquibaseCommandLine.java:353)
	at liquibase.Scope.child(Scope.java:189)
	at liquibase.Scope.child(Scope.java:165)
	at liquibase.integration.commandline.LiquibaseCommandLine.execute(LiquibaseCommandLine.java:318)
	at liquibase.integration.commandline.LiquibaseCommandLine.main(LiquibaseCommandLine.java:84)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at liquibase.integration.commandline.LiquibaseLauncher.main(LiquibaseLauncher.java:107)
Caused by: liquibase.exception.ValidationFailedException: Validation Failed:
     1 changesets check sum
          update/services/messaging/messaging_tables.xml::1559570706::mw185149 was: 8:e1aa84a7a58273fe5908ad86a589612d but is now: 8:dce471b939276f17937a806ee31e276b

	at liquibase.changelog.DatabaseChangeLog.validate(DatabaseChangeLog.java:339)
	at liquibase.Liquibase.lambda$update$1(Liquibase.java:229)
	at liquibase.Scope.lambda$child$0(Scope.java:180)
	at liquibase.Scope.child(Scope.java:189)
	at liquibase.Scope.child(Scope.java:179)
	at liquibase.Scope.child(Scope.java:158)
	at liquibase.Liquibase.runInScope(Liquibase.java:2414)
	at liquibase.Liquibase.update(Liquibase.java:209)
	at liquibase.Liquibase.update(Liquibase.java:195)
	at liquibase.integration.commandline.Main.doMigration(Main.java:1845)
	at liquibase.integration.commandline.Main$1.lambda$run$0(Main.java:398)
	at liquibase.Scope.lambda$child$0(Scope.java:180)
	at liquibase.Scope.child(Scope.java:189)
	at liquibase.Scope.child(Scope.java:179)
	at liquibase.Scope.child(Scope.java:158)
	at liquibase.integration.commandline.Main$1.run(Main.java:397)
	at liquibase.integration.commandline.Main$1.run(Main.java:219)
	at liquibase.Scope.child(Scope.java:189)
	at liquibase.Scope.child(Scope.java:165)
	at liquibase.integration.commandline.Main.run(Main.java:219)
	at liquibase.command.AbstractCliWrapperCommandStep.run(AbstractCliWrapperCommandStep.java:33)
	at liquibase.command.CommandScope.execute(CommandScope.java:173)
	... 20 more
`

As you can see, the issue seems to be related with the computed checksum for the first replace subtag (which are different regardless nothing was changed).


`[2022-11-11 06:34:49] FINE [liquibase.util] Computed checksum for replace:[
    applicableDbms=[
        oracle,
        mssql
    ]
    applyToRollback="false"
    contexts="non-iot"
    replace=", CONSTRAINT PK_cs_guaranteed_delivery PRIMARY KEY (internal_id)"
    with=""
] as **d0e8ec5e3b5380571154dfb483e6080a**

[2022-11-11 07:45:28] FINE [liquibase.util] Computed checksum for replace:[
    applicableDbms=[
        oracle,
        mssql
    ]
    applyToRollback="false"
    contextFilter="non-iot"
    replace=", CONSTRAINT PK_cs_guaranteed_delivery PRIMARY KEY (internal_id)"
    with=""
] as **f60940602dcd93a68b2b4a93da27ac35**
`

Thanks for your time and help on this. Looking forward to hearing from you.
Regards,
Italo.

**Comment by nvoxland** [comment]
Thanks for the logs, they do point out where the problem is. It's if you have a `<modifySql>` tag with a `context` attribute on elements within it. We had changed that attribute from `context` to `contextFilter` to improve usability and support backwards compatibility when specifying just `context`, BUT the attribute name is used in the checksum string and so we get a difference there. 

We'll have to find a way to preserve the checksum even with the new canonical attribute name.

**Comment by filipelautert** [comment]
Hello @ic185037 - PR https://github.com/liquibase/liquibase/pull/3530 aims to fix the issue you raised. By reviewing the code it is clear that md5 generation for some sql functions changed on Liquibase 4.16 . When this gets merged it will be back to how it was before - but then MD5 generated from 4.16 until PR release will be broken (probably 4.18.1 or 4.19). In that case, the fix is what you did: clear md5sum before running liquibase or use "clear-checksums" command.

**Comment by filipelautert** [comment]
Hello @doakd , @CGarces - can you confirm if the problem happens everytime you execute update on the same liquibase version, or does it happens just when upgrading the version?
If it happens everytime, could you just share a small sample of what you think may be causing the problem? thanks

**Comment by doakd** [comment]
It happens everytime "update" is run.  no "clear-checksums" or clear md5sum is being run.  The only examples I have are what I pasted above.

**Comment by ic185037** [comment]
> Hello @ic185037 - PR #3530 aims to fix the issue you raised. By reviewing the code it is clear that md5 generation for some sql functions changed on Liquibase 4.16 . When this gets merged it will be back to how it was before - but then MD5 generated from 4.16 until PR release will be broken (probably 4.18.1 or 4.19). In that case, the fix is what you did: clear md5sum before running liquibase or use "clear-checksums" command.

Hi @filipelautert 
Thank you for your reply and for your time and effort on this fix. I sincerely appreciate it. Do you have an estimated release date for 4.18.1 or 4.19 releases?
Thanks

**Comment by filipelautert** [comment]
@ic185037 we try to have a release each 4~5 weeks, so maybe we may release it around Xmas/new year.

**Comment by filipelautert** [comment]
@doakd sorry, I'm not able to reproduce the issue. But @nvoxland found out a different bug that _may_ be related to it: https://github.com/liquibase/liquibase/issues/3549 .

**Comment by ic185037** [comment]
> @ic185037 we try to have a release each 4~5 weeks, so maybe we may release it around Xmas/new year.

Hi @filipelautert @nvoxland @XDelphiGrl @MalloD12 

Sorry to bother you on this issue again.

I cannot see this issue fixed in the latest release (4.19.0). From the latest Release Notes, I cannot see neither PR #3269 nor PR #3530 but they seems to be Closed. Was this issue fixed as part of a different PR included in 4.19.0?

Thanks and Regards.

**Comment by filipelautert** [comment]
Hi @ic185037 - at first we were implementing a fix specific to one of those issues. But as we tried to fix it we found more and more problems, so we decided to work on a bigger Pr that will fix it for once. 
As it will be a big change we will need to bump the md5sum algorithm version (that will trigger a md5 recalculation for everyone), so we are trying to add all the possible fixes in this pr. We may merge it to master branch next week and hopefully it will make to next release.

Sorry for the delay on fixing this. The PR including the fixes is https://github.com/liquibase/liquibase/pull/3616 .
